Add Troubleshooting guide (#119)
submariner-io/submariner-website: src/content/troubleshooting/_index.en.md

+++
title = "Troubleshooting Guide"
date = 2020-05-04T19:01:14+05:30
weight = 5
pre = "<b>4. </b>"
+++

## Overview

You have followed steps in [Deployment](../deployment) but something has gone wrong. You're not sure what and how to fix it, or what information to collect to raise an issue. Welcome to the Submariner troubleshooting guide where we will help you get your deployment working again.

Basic familiarity with the Submariner components and architecture will be helpful when troubleshooting so please review the [Architecture](../architecture) section.

The guide has been broken into different sections for easy navigation.

### Pre-requisite
Before we begin troubleshooting, run `subctl version` to obtain which version of the Submariner components you are running.

> Run `kubectl get services | grep <service-name>` to get information about the service you're trying to access. This will provide you with the Service *Name*, *Namespace* and *ServiceIP*. If **Globalnet** is enabled, you will also need the *globalIp* of the service by running

Reviewer: Pls include that "kubectl get services ..." should be queried in the respective namespace where the service is deployed.

``` kubectl get service <service-name> -o jsonpath='{.metadata.annotations.submariner\.io/globalIp}' ```

<!---
### Deployment Issues
This section will contain information about common deployment issues you can run into.

#### TBD

### Connectivity Issues
Deployment went through successfully but services/pods on one cluster are unable to connect to services on another clusters. This can be due to multuple factors - IP Sec tunnels, IP Table rules, Gateway pods, Routeagent etc.

#### IPSec tunnel not created between clusters
TBD

#### IPSEc tunnel is not up between clusters
TBD

#### None of pods/services able to connect to remote service
TBD
##### Without Globalnet
TBD
##### With Globalnet
TBD

#### Pods on non-gateway nodes not able to connect to remote service
TBD
##### Without Globalnet
TBD
##### With Globalnet
TBD

-->

### Service Discovery Issues
If you are able to connect to remote service by using ServiceIP or globalIp, but not by service name, it is a Service Discovery Issue.

#### Service Discovery not working

> This is good time to familiarize yourself with [ServiceDiscovery Architecture](../architecture/service-discovery/) if you haven't already.

Reviewer: nit: please give space between ServiceDiscovery (aka Service Discovery)

##### Check CoreDNS Configuration
Submariner configures CoreDNS deployment to enable `lighthouse` plugin. Since CoreDNS is the first point of entry for any DNS queries, that is where we will start first. 

First we check if we are running the correct image. We need CoreDNS image with ligthouse enabled.

> ```kubectl -n kube-system kube-dns describe deployment```

Reviewer: In OpenShift, I dont see the kube-dns deployment. Can you also include a note about equivalent command for OCP?

Author: Made a generic note to cover all non-vanilla deployments.

Reviewer: I tried the command in OCP and it does not show any dns deployments in openshift-dns namespace.

[sgaddam@localhost aws-sgaddam]$ kubectl -n openshift-dns get deployment
No resources found.
[sgaddam@localhost aws-sgaddam]$

+++
title = "Troubleshooting Guide"
date = 2020-05-04T19:01:14+05:30
weight = 5
pre = "<b>4. </b>"
+++

## Overview

You have followed steps in [Deployment](../deployment) but something has gone wrong. You're not sure what and how to fix it, or what information to collect to raise an issue. Welcome to the Submariner troubleshooting guide where we will help you get your deployment working again.

Basic familiarity with the Submariner components and architecture will be helpful when troubleshooting so please review the [Architecture](../architecture) section.

The guide has been broken into different sections for easy navigation.

### Pre-requisite
Before we begin troubleshooting, run `subctl version` to obtain which version of the Submariner components you are running.

> Get information about the service you're trying to access.

Reviewer 2: This line can now be removed as its redundant with L21.

##### Check submariner-lighthouse-agent
Next we check if `submariner-lighthouse-agent` is running correctly or not. Run `kubectl -n submariner-operator get pods submariner-lighthouse-agent` and check status of pods.

If status is error with `ImagePullBackOff`, run `kubectl -n submariner-operator describe deployment submariner-lighthouse-agent` and check if `Image` is set correctly to `quay.io/submariner/lighthouse-agent:<version>`. If it is and still getting the same error, raise an issue [here](https://github.com/submariner-io/lighthouse/issues) or ping us on the community slack channel.

If status is just error, run `kubectl -n submariner-operator get pods` to get the pod name for lighthouse agent. Then run `kubectl -n submariner-operator logs <lighthouse-agent-pod-name>` to get the logs. See if there are any errors in the log. If yes, raise an [issue](https://github.com/submariner-io/lighthouse/issues) with log contents, or you can continue reading through document to troubleshoot this further.

If there are no errors, grep the logs for service name that you're trying to query, we may need them later for raising an issue.

##### Check multiclusterservice CRs

> If the steps above did not indicate and issue, next we check if the Multiclusterservice resources were properly created for the service you're trying to access. The format of Multiclusterservice resources's name is as follows:

Reviewer 2: ```suggestion
If the steps above did not indicate an issue, next we check if the Multiclusterservice resources were properly created for the service you're trying to access. The format of a Multiclusterservice resources's name is as follows:
```

`<service-name>-<service-namespace>-<cluster-id>`

> Run `kubectl get multiclusterservices --all-namespaces |grep <your-service-name>` on the broker cluster to chheck if a resource was created for your service. If not, the check the Lighthouse Agent logs on the cluster where service was created and look for any error or warning messages indicating a failure to create the Multiclusterservice resource for your service. The most common error is `Forbidden` if the RBAC wasn't configured correctly. Depending on the deployment method used, 'subctl' or 'helm', it should've been done for you. Create an [issue](https://github.com/submariner-io/lighthouse/issues) with relevant log entries.

Reviewer 2: ```suggestion
Run `kubectl get multiclusterservices --all-namespaces |grep <your-service-name>` on the broker cluster to check if a resource was created for your service. If not, the check the Lighthouse Agent logs on the cluster where service was created and look for any error or warning messages indicating a failure to create the Multiclusterservice resource for your service. The most common error is `Forbidden` if the RBAC wasn't configured correctly. Depending on the deployment method used, 'subctl' or 'helm', it should've been done for you. Create an [issue](https://github.com/submariner-io/lighthouse/issues) with relevant log entries.
```

+++

> title = "Troubleshooting Guide"

Reviewer 3: Nit: for consistency, let's also change "Quickstart guide" to "Quickstart Guide" and "Online resources" to "Online Resources". I will submit a PR for that.

date = 2020-05-04T19:01:14+05:30
weight = 5

> pre = "<b>1. </b>"

Reviewer 3: This guide should not be number 1 (the first one) we show on the menu. I suggest to make it number 4. It also means that other pages would need to be updated so that we have the correct numbering:

1. Quickstart guide
2. Deployment
3. Architecture
4. Troubleshooting Guide
5. Releases
6. Online resources
7. Roadmap
8. Contributing

Run `subctl version` to get version information.

Get information about the service you're trying to access.

`kubectl get services |grep <service-name>` will provide you with Service *Name*, *Namespace* and *ServiceIP*. If you enabled **Globalnet** you will also need the *GlobalnetIP* of the service by running

``` kubectl get service <service-name> -o jsonpath='{.metadata.annotations.submariner\.io/globalIp}' ```

<!---
### Deployment Issues
This section will contain information about common deployment issues you can run into.

#### TBD

### Connectivity Issues
Deployment went through successfully but services/pods on one cluster are unable to connect to services on another clusters. This can be due to multuple factors - IP Sec tunnels, IP Table rules, Gateway pods, Routeagent etc.

#### IPSec tunnel not created between clusters
TBD

#### IPSEc tunnel is not up between clusters
TBD

#### None of pods/services able to connect to remote service
TBD
##### Without Globalnet
TBD
##### With Globalnet
TBD

#### Pods on non-gateway nodes not able to connect to remote service
TBD
##### Without Globalnet
TBD
##### With Globalnet
TBD

-->

### Service Discovery Issues

> If you are able to connect to remote service by using ServiceIP or GlobalnetIP, but not by service name, it is a Service Discovery Issue.

Reviewer: Same here for GlobalnetIP

#### Service Discovery not working
This is good time to familiarize yourself with [Lighthouse Architecture](../architecture/components/lighthouse) if you haven't already.

> ##### Check CoreDNS Configuration

Reviewer 4: Maybe a section to check operator CR too , if it is an operator based deployment?

Author: That would go in Deployment section of the guide.

> Submariner configures CoreDNS deployment to enable `lighthouse` plugin. Since CoreDNS is the first point of entry for any DNS queries, that is where we will start first. 

Author: There is no "installation". Plugin is already part of coredns, we only enable it.

First we check if we are running the correct image. We need CoreDNS image with ligthouse enabled.

```kubectl -n kube-system kube-dns describe deployment```

Make sure that `Image` is set to `quay.io/submariner/lighthouse-coredns:<version>` and version is correct. If not, change it to use the Lighthouse image.

> If image is correct, next we need to check if `lighthouse` plugin is enabled on the CoreDNS in the source cluster i.e. the cluster making the query.

Author: We use `lighthouse` because keyword in config is `lighthouse`

If there are no errors, grep the logs for service name that you're trying to query, we may need them later for raising an issue.

##### Check multiclusterservice CRs
If all this was fine, next we need to check if `multiclusterservices` CRs were created correctly for the service you're trying to access. The format of `multiclusterservice` object's name is as follows:

`<service-name>-<service-namespace>-<cluster-id>`

Run `kubectl get multiclusterservices --all-namespaces |grep <your-service-name>` on the broker cluster to see if CR was created for your service or not. If not, check the Lighthouse Agent logs on the cluster where service was created, there should be some error or warning messages about being unable to create `multiclusterservice` object for your service. Most common error can be `Forbidden` if the RBAC wasn't configured correctly. Depending on deployment used, *subctl* or *helm* it should've been done for you. Create an [issue](https://github.com/submariner-io/lighthouse/issues) with relevant log entries.

> If the CR was created correctly on broker cluster, next step is to check for it on the cluster where you're trying to access the service. Follow the same steps as earlier to get list of `multiclusterservices` objects created and if your service name shows up in it or not. If not, again check the logs of Lighthouse Agent but this time on the cluster where you are trying to access the service. Again, it will likely be an issue with RBAC so you follow the same steps and create an [issue](https://github.com/submariner-io/lighthouse/issues) with relevant log entries.

Reviewer 2: ```suggestion
If the Multiclusterservice resource was created correctly on the broker cluster, the next step is to check if it exists on the cluster where you're trying to access the service. Follow the same steps as earlier to get the list of the Multiclusterservice resources and check if the Multiclusterservice for your service exists. If not, check the logs of the Lighthouse Agent on the cluster where you are trying to access the service. As described earlier, it will most commonly be an issue with RBAC otherwise create an [issue](https://github.com/submariner-io/lighthouse/issues) with relevant log entries.
```

> If the CR was created correctly, run `kubectl -n submariner-operator describe multiclusterservice <your-multiclusterservice-name>` and check if it has correct `ClusterID` and `ServiceIP`.

Reviewer 2: ```suggestion
If the Multiclusterservice resource was created properly on the cluster, run `kubectl -n submariner-operator describe multiclusterservice <your-multiclusterservice-name>` and check if it has the correct `ClusterID` and `ServiceIP`:
```

```
Name:         nginx-demo-default-cluster2
Namespace:    submariner-operator
Labels:       submariner-io/clusterID=cluster2
Annotations:  <none>
API Version:  lighthouse.submariner.io/v1
Kind:         MultiClusterService
Metadata:
  Creation Timestamp:  2020-04-16T15:40:36Z
  Generation:          1
  Resource Version:    6847
  Self Link:           /apis/lighthouse.submariner.io/v1/namespaces/submariner-operator/multiclusterservices/nginx-demo-ns-default-src-cluster2
  UID:                 9db7759c-7ff8-11ea-85a3-0242ac110006
Spec:
  Cluster Service Info:
    Cluster Domain:  
    Cluster ID:      cluster2       ==========> ClusterID of cluster where service is running
    Service IP:      100.92.243.156 ==========> ServiceIP or GlobalIP of service you're trying to access
Events:              <none>
```

> If not correct, you can manually edit the `multiclusterservices` CR to the correct IP as a workaround and create an [issue](https://github.com/submariner-io/lighthouse/issues) with relevant information.

Reviewer 2: ```suggestion
If the data is not correct, you can manually edit the Multiclusterservice resource to set the correct IP as a workaround and create an [issue](https://github.com/submariner-io/lighthouse/issues) with relevant information.
```

+++
title = "Troubleshooting Guide"
date = 2020-05-04T19:01:14+05:30
weight = 5
pre = "<b>1. </b>"
+++

## Overview

> You have folowed steps in [Deployment](../deployment) but something has gone wrong. You're not sure what and how to fix it, or what information to collect to raise an issue. Welcome to the Submariner troubleshooting guide where we will help you get your deployment working again.

Reviewer 5: ```suggestion
You have followed steps in [Deployment](../deployment) but something has gone wrong. You're not sure what and how to fix it, or what information to collect to raise an issue. Welcome to the Submariner troubleshooting guide where we will help you get your deployment working again.
```

To recap, Submariner consists of two main core components (broker and submariner) and optional components (lighthouse), detailed in [Architecture](../architecture) section. Basic familiarity with the architecture will be helpful when troubleshooting.

The guide has been broken into different sections for easy navigation.

### Pre-requisite
Before we begin troubleshooting, it would be good to know which version of submariner and its components you are running.

{{% notice info %}}

> TBD: Add information on how to find out which version users are running. Is `subctl version` enough?

Reviewer 6: subctl version only shows the version of the subctl binary, but that'd be a cool improvement.

We can look at the operator status, may be we can add also that information into subctl "version" -> looks the operator status, and returns those details too?

Author: I'll raise an issue to add more useful information to `subctl version`

{{% /notice %}}

Get information about the service you're trying to access.

> `kubectl get services |grep <service-name` will provide you with Service *Name*, *Namespace* and *ServiceIP*. If you enabled **Globalnet** you will also need the *GlobalnetIP* of the service by running

Reviewer 6: remove "<" 
```suggestion
`kubectl get services |grep service-name` will provide you with Service *Name*, *Namespace* and *ServiceIP*. If you enabled **Globalnet** you will also need the *GlobalnetIP* of the service by running
```

Reviewer 5: I would suggest to add > instead. service-name is a place holder here.

Author: @Jaanki is correct, I missed adding ">".

``` kubectl get service <service-name> -o jsonpath='{.metadata.annotations.submariner\.io/globalIp}' ```

### Deployment Issues
This section will contain information about common deployment issues you can run into.

#### TBD

### Connectivity Issues

> Deployment went through successfully but services/pods on one cluster are unable to connect to services in another clusters. This can be due to multuple factors - IP Sec tunnels, IP Table rules, Gateway pods, Routeagent etc.

Reviewer 5: ```suggestion
Deployment went through successfully but services/pods on one cluster are unable to connect to services in another clusters. This can be due to multuple factors - IP Sec tunnels, IP Table rules, Gateway pods, Routeagent, propoer ports not being opened etc.
```

Reviewer 5: ALso maybe a line saying that internet connectivity is needed when connecting private clusters like need of a PublicIP in case of AWS

Author: I think etc. covers all that :D

> Deployment went through successfully but services/pods on one cluster are unable to connect to services in another clusters. This can be due to multuple factors - IP Sec tunnels, IP Table rules, Gateway pods, Routeagent etc.

Reviewer 5: ```suggestion
Deployment went through successfully but services/pods on one cluster are unable to connect to services on another clusters. This can be due to multuple factors - IP Sec tunnels, IP Table rules, Gateway pods, Routeagent etc.
```

#### IPSec tunnel not created between clusters

> TBD

Reviewer 6: Now we can show this up with the:

kubectl get gateways -n submariner-operator

and kubectl describe gateway <hostname> , it will have a connection list.

note to myself: We can add those details to subctl too in a more visual way.

Reviewer 6: The reasons could be: 

* Wrong PSK
* Firewall issues

Reviewer 6: May be ... we can comment the sections we don't have content for... and have them populated on later PRs? 

you can use <!--   --> to comment specific areas of in the markdown I believe.

Author: commented out incomplete sections

> #### IPSEc tunnel is not up between clusters

Reviewer 5: Check ports are open on nodes

Reviewer 6: We should do that in a later PR, I asked vthapar to keep those sections commented , so they don't render on the site, but we can keep improving as individual PRs.

TBD

#### None of pods/services able to connect to remote service
TBD
##### Without Globalnet
TBD
##### With Globalnet
TBD

#### Pods on non-gateway nodes not able to connect to remote service
TBD
##### Without Globalnet
TBD
##### With Globalnet
TBD

### Service Discovery Issues
If you are able to connect to remote service by using ServiceIP or GlobalnetIP, but not by service name, it is a Service Discovery Issue.

#### Service Discovery not working
This is good time to familiarize yourself with [Lighthouse Architecture](../architecture/components/lighthouse) if you haven't already.

##### Check CoreDNS Configuration

> Submariner configures CoreDNS deployment to enable `lighthouse` plugin. Since CoreDNS is the first point of entry for any DNS queries, that is where we will start first. 

Reviewer 6: This is awesome! :)

If image is correct, next we need to check if `lighthouse` plugin is enabled on the CoreDNS in the source cluster i.e. the cluster making the query.

```kubectl describe configmap coredns```

In the output look for something like this:

```
        kubernetes cluster2.local in-addr.arpa ip6.arpa {
           pods insecure
           upstream
           #fallthrough in-addr.arpa ip6.arpa
           fallthrough     =====> This tells kubernetes to go to next plugin if query fails locally
        }
        lighthouse { ====> Lighthouse plugin will catch any queries that local kubernetes can't resolve
           fallthrough =====> proceed as usual if lighthouse can't resolve
        }
```
If the entries highlighted above are missing, it means CoreDNS wasn't configured with lighthouse. It can be easily added by running `kubectl edit configmap coredns` and making the changes manually. You may need to repeat this step on every cluster.

##### Check submariner-lighthouse-agent
Next we check if `submariner-lighthouse-agent` is running correctly or not. Run `kubectl -n submariner-operator get pods submariner-lighthouse-agent` and check status of pods.

> If status is error with `ImagePullBackOff`, run `kubectl -n submariner-operator describe deployment submariner-lighthouse-agent` and check if `Image` is set correctly to `quay.io/submariner/lighthouse-agent:<version>`. If it is and still getting the same error, raise an issue as image has likely gone missing from quay.io

Reviewer 6: ```suggestion
If status is error with `ImagePullBackOff`, run `kubectl -n submariner-operator describe deployment submariner-lighthouse-agent` and check if `Image` is set correctly to `quay.io/submariner/lighthouse-agent:<version>`. If it is and still getting the same error, raise an issue [here](https://github.com/submariner-io/lighthouse/issues) or ping us on the community slack channel.
```

Reviewer 5: > 

Also add link to community slack channel

Author: @Jaanki Good idea, but think we need a "Contact" section on website and link to it.

If there are no errors, grep the logs for service name that you're trying to query, we may need them later for raising an issue.

##### Check multiclusterservice CRs
If all this was fine, next we need to check if `multiclusterservices` CRs were created correctly for the service you're trying to access. The format of `multiclusterservice` object's name is as follows:

`<service-name>-<service-namespace>-<cluster-id>`

Run `kubectl get multiclusterservices --all-namespaces |grep <your-service-name>` on the broker cluster to see if CR was created for your service or not. If not, check the Lighthouse Agent logs on the cluster where service was created, there should be some error or warning messages about being unable to create `multiclusterservice` object for your service. Most common error can be `Forbidden` if the RBAC wasn't configured correctly. Depending on deployment used, *subctl* or *helm* it should've been done for you. Create an [issue](https://github.com/submariner-io/lighthouse/issues) with relevant log entries.

> If the CR was created correctly on broker cluster, next step is to check for it on the cluster where you're trying to access the service. Follow the same steps as earlier to get list of `multiclusterservices` objects created and if your service name shows up in it or not. If not, again check the logs of Lighthouse Agent but this time on the cluster where you are trying to access the service. Again, it will likely be an issue with RBAC and you follow the same steps, create an Create an [issue](https://github.com/submariner-io/lighthouse/issues) with relevant log entries.

Reviewer 5: ```suggestion
If the CR was created correctly on broker cluster, next step is to check for it on the cluster where you're trying to access the service. Follow the same steps as earlier to get list of `multiclusterservices` objects created and if your service name shows up in it or not. If not, again check the logs of Lighthouse Agent but this time on the cluster where you are trying to access the service. Again, it will likely be an issue with RBAC and you follow the same steps, create an [issue](https://github.com/submariner-io/lighthouse/issues) with relevant log entries.
```

> If the CR was created correctly, run `kubectl -n submariner-operator describe multiclusterservice <your-mcs-name>` and check if it has correct `ClusterID` and `ServiceIP`.

Reviewer 5: mcs appears for the first time. Should we somewhere expand the abbreviation?

Author: done

```
Name:         nginx-demo-default-cluster2
Namespace:    submariner-operator
Labels:       submariner-io/clusterID=cluster2
Annotations:  <none>
API Version:  lighthouse.submariner.io/v1
Kind:         MultiClusterService
Metadata:
  Creation Timestamp:  2020-04-16T15:40:36Z
  Generation:          1
  Resource Version:    6847
  Self Link:           /apis/lighthouse.submariner.io/v1/namespaces/submariner-operator/multiclusterservices/nginx-demo-ns-default-src-cluster2
  UID:                 9db7759c-7ff8-11ea-85a3-0242ac110006
Spec:
  Cluster Service Info:
    Cluster Domain:  
    Cluster ID:      cluster2       ==========> ClusterID of cluster where service is running
    Service IP:      100.92.243.156 ==========> ServiceIP or GlobalIP of service you're trying to access
Events:              <none>
```

> If not correct, you can manually edit the `multiclusterservices` CR to the correct IP as a workaround and Create an [issue](https://github.com/submariner-io/lighthouse/issues) with relevant information.

Reviewer 5: ```suggestion
If not correct, you can manually edit the `multiclusterservices` CR to the correct IP as a workaround and create an [issue](https://github.com/submariner-io/lighthouse/issues) with relevant information.
```

+++
title = "Troubleshooting Guide"
date = 2020-05-04T19:01:14+05:30

> weight = 20

Reviewer 2: This weight will be correct with https://github.com/submariner-io/submariner-website/pull/134 - should rebase on top of that (#134 is ready too merge anyay).

Author: Done. Check the order. I'm going with weight = order*5

### Pre-requisite
Before we begin troubleshooting, run `subctl version` to obtain which version of the Submariner components you are running.

> Run `kubectl get services -n <service-namespace> | grep <service-name>` to get information about the service you're trying to access. This will provide you with the Service *Name*, *Namespace* and *ServiceIP*. If **Globalnet** is enabled, you will also need the *globalIp* of the service by running

Reviewer: nit: Our docs use a convention of "GlobalNet". It would be nice to follow the same convention everywhere.

#### TBD

### Connectivity Issues

> Submariner deployment completed successfully but Services/Pods on one cluster are unable to connect to Services on another cluster. This can be due to multiple factors outlined in the following sections.

Reviewer: This section is not getting rendered in the final output. Can you take a look.

Author: That is by design, refer @mangelajo comment. It is commented out and idea is others will add content to it and enable it. Focus of this PR is lighthouse.

If you are able to connect to remote service by using ServiceIP or globalIp, but not by service name, it is a Service Discovery Issue.

#### Service Discovery not working
This is good time to familiarize yourself with [Service Discovery Architecture](../architecture/service-discovery/) if you haven't already.

##### Check CoreDNS Configuration
Submariner configures CoreDNS deployment to enable `lighthouse` plugin. Since CoreDNS is the first point of entry for any DNS queries, that is where we will start first. 

First we check if we are running the CoreDNS image with the Lighthouse plugin enabled.

```kubectl -n kube-system kube-dns describe deployment```

{{% notice info %}}

> For Openshift use ```kubectl -n openshift-dns dns describe deployment```

Reviewer: nit: OpenShift

> Some deployments may use different name and namespace for DNS service e.g. `coredns`, `core-dns` etc.

Reviewer: This is not getting rendered properly in the final output. 

![9](https://user-images.githubusercontent.com/6574656/81540749-b5ec7380-938f-11ea-965c-86309c3498d6.jpg)

Author: Should work correctly now
